File: 2021/15/solution.py
import os
file1 = open(os.path.join(os.path.dirname(os.path.abspath(__file__)), 'input.txt'),'r')
lines = file1.readlines()

import heapq
import math
from collections import defaultdict

risks = [list(map(int,list(line.strip()))) for line in lines]

def findMinRisk(grid):
  visited = set()

  h = len(grid)
  w = len(grid[0])

  start = (0, 0)
  dest = (h - 1, w - 1)

  minRisk = defaultdict(lambda: math.inf)
  minRisk[start] = 0

  q = [(0, start)]

  while len(q):
    risk, pos = heapq.heappop(q)
    if pos == dest:
      return risk
    if pos in visited:
      continue
    visited.add(pos)
    r, c = pos
    if r > 0 and (r - 1, c) not in visited:
      newRisk = risk + grid[r - 1][c]
      if newRisk < minRisk[(r - 1, c)]:
        minRisk[(r - 1, c)] = newRisk
        heapq.heappush(q, (newRisk, (r - 1, c)))
    if c > 0 and (r, c - 1) not in visited:
      newRisk = risk + grid[r][c - 1]
      if newRisk < minRisk[(r, c - 1)]:
        minRisk[(r, c - 1)] = newRisk
        heapq.heappush(q, (newRisk, (r, c - 1)))
    if r < h - 1 and (r + 1, c) not in visited:
      newRisk = risk + grid[r + 1][c]
      if newRisk < minRisk[(r + 1, c)]:
        minRisk[(r + 1, c)] = newRisk
        heapq.heappush(q, (newRisk, (r + 1, c)))
    if c < w - 1 and (r, c + 1) not in visited:
      newRisk = risk + grid[r][c + 1]
      if newRisk < minRisk[(r, c + 1)]:
        minRisk[(r, c + 1)] = newRisk
        heapq.heappush(q, (newRisk, (r, c + 1)))

print(findMinRisk(risks))

h = len(risks)
w = len(risks[0])
risks2 = [[0] * 5 * w for _ in range(5 * h)]

for r in range(5):
  for c in range(5):
    for i in range(h):
      for j in range(w):
        tmp = risks[i][j] + r + c
        if tmp > 9:
          tmp -= 9
        risks2[h * r + i][w * c + j] = tmp

print(findMinRisk(risks2))

File: 2021/15/solution.js
const fs = require("fs");
const path = require("path");
const MinHeap = require("../../utils/MinHeap");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split("\n")
  .map((a) => a.split("").map(Number));

function findMinRisk(input) {
  const turnDirections = [
    [0, 1],
    [1, 0],
    [-1, 0],
    [0, -1],
  ];

  const rMax = input.length - 1;
  const cMax = input[0].length - 1;

  const queue = [[0, 0, 0]];
  const visited = new Map();

  while (queue.length) {
    const [risk, r, c] = MinHeap.pop(queue);
    if (r === rMax && c === cMax) {
      return risk;
    }
    for (const [dr, dc] of turnDirections) {
      const r2 = r + dr;
      const c2 = c + dc;
      if (r2 < 0 || c2 < 0 || r2 > rMax || c2 > cMax) continue;
      const risk2 = risk + input[r2][c2];
      if (visited.has(`${r2},${c2}`) && visited.get(`${r2},${c2}`) <= risk2)
        continue;
      visited.set(`${r2},${c2}`, risk2);
      MinHeap.push(queue, [risk + input[r2][c2], r2, c2]);
    }
  }
}

function expandInput(input) {
  const rows = input.length;
  const cols = input[0].length;
  for (let i = 0; i < 4; i++) {
    for (let r = rows * i; r < rows * (i + 1); r++) {
      const newRow = [];
      for (let c = 0; c < cols; c++) {
        let val = input[r][c] + 1;
        if (val > 9) val = 1;
        newRow.push(val);
      }
      input.push(newRow);
    }
  }
  for (let i = 0; i < 4; i++) {
    for (let r = 0; r < input.length; r++) {
      for (let c = cols * i; c < cols * (i + 1); c++) {
        let val = input[r][c] + 1;
        if (val > 9) val = 1;
        input[r].push(val);
      }
    }
  }

  return input;
}

console.log(findMinRisk(input));
console.log(findMinRisk(expandInput(input)));
